In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df=pd.read_csv(r"D:\Data Analysis\Projects\Loan Analyser eda\data\Cleaned_Dataset.csv")
display(df)
display(df.head())
display(df.tail())
display(df.sample())
display(df.isnull().sum())
display(df.duplicated())
display(df.info())
display(df.describe())
display(df.dtypes)
display(df.shape)

In [ ]:
display(df.info())



# Let's classify the columns first

# 1. Numerical columns
# age
# monthly_income_inr
# existing_loans_count
# loan_amount_requested_inr
# existing_emi_monthly_inr
# credit_utilization_ratio_pct
# interest_rate_offered_pct
# bank_account_vintage_years
# num_credit_inquiries_last_6m
# late_payments_last_12m
# loan_tenure_months
# cibil_score
# default_risk_score


# 2. Categorical columns
# gender
# marital_status
# city_tier
# state
# employment_type
# loan_purpose
# collateral_provided
# default_flag


# 3. Identifier column
# applicant_id

#univariate analysis on Numerical colums


In [ ]:
#age_column 
display(df["age"].describe())
#mean-> 35.31
#median-> 35
#(max,min)-> (65,21)
#(q1,q3)->(29,41)


#creating visuals 
df["age"].plot(kind="hist",bins=10)
plt.show()
#histogram


df["age"].plot(kind="density")
plt.show()
#density plot

#checking skewness
display(df["age"].skew()) #0.265

#boxplot
df["age"].plot(kind="box")
plt.show()
#outliners are present in the age column

#Summary of univariate analysis on Age column
#1.> the avg age of customer is 35.31 
#2.> median age  of the customers is  age 35
#3.> min age of customer is 21 ,while the max age is 65
#4.> by looking at histogram we can say that as we are inc age after range of 30-40 the no of customer is decreasing
#5.> there is a negilible skewness in the columns
#6.> by looking at box plot a outliner is present after age 65
#7.> but the age coloums has acceptable values (so we can neglecate it)

In [ ]:
#univariate analysis on montly_icome_inr coloumn
display(df["monthly_income_inr"].describe())
#mean-> 54784.481
#(max,min)->(727000,12000)
#(q1,q2)-> (29400,68700)
#median-> 45100
df["monthly_income_inr"].plot(kind="hist")
plt.show()
df["monthly_income_inr"].plot(kind="density")
plt.show()
display(df["monthly_income_inr"].skew()) # 2.742
df["monthly_income_inr"].plot(kind="box")
plt.show()



# Summary of univariate analysis on monthly_income_inr column

# 1. The average monthly income of customers is ₹54,784.48.

# 2. The median monthly income of customers is ₹45,100.

# 3. The minimum monthly income is ₹12,000,
#    while the maximum monthly income is ₹727,000.

# 4. The distribution is positively skewed.

# 5. The boxplot shows the presence of multiple high-income outliers.

# 6. The maximum income is significantly higher than the majority
#    of customer incomes.

# 7. However, high income does not necessarily mean invalid data,
#    so the outliers will be retained.

In [ ]:
#univariate analysis on existing_loans_count
display(df["existing_loans_count"].describe())
#mean-> 1.104
#(max,min)-(0,6)
#median-1
#(q1,q3)->(0,2)
df["existing_loans_count"].plot(kind="hist")
plt.show()
df["existing_loans_count"].plot(kind="density")
plt.show()
display(df["existing_loans_count"].skew()) #0.95
df["existing_loans_count"].plot(kind="box")
plt.show()




In [ ]:
display(df[df["existing_loans_count"]==6].shape)
#22 customer has 6 existing loan counts on it
display(df[df["existing_loans_count"]>2].shape)
#2442 customer has more than 2 existing loan counts on it


#summary of univariate Analysis
1.> the avg existing loan count of customer is approximately 1 (1.104)
2.> the median of loan count of customer is 1
3.> by looking the histogram we can coclude that as the loan count increases the no of customer descreases
4.> this column has positive skewness
5.>  The maximum existing loan count is 6, with 22 customers having 6 existing loans.
6.> Even if higher loan counts are statistically identified as outliers by box plot
  they do not necessarily represent invalid data, as having multiple loans is possible.




In [ ]:
#univariate analysis on existing emi montly inr
display(df["existing_emi_monthly_inr"].describe())
#mean-> 8437.618
#(max,min)-(202300,0)
#median-5700
#(q1,q3)->(0,12400)
df["existing_emi_monthly_inr"].plot(kind="hist")
plt.show()
df["existing_emi_monthly_inr"].plot(kind="density")
plt.show()
display(df["existing_emi_monthly_inr"].skew()) #2.285
df["existing_emi_monthly_inr"].plot(kind="box")
plt.show()

display(df[df["existing_emi_monthly_inr"]>200000]) #6102
#0ne customer but this one customer has also highest montly income 

display(df[df["existing_emi_monthly_inr"]>100000]) #15 customers
display(df[df["existing_emi_monthly_inr"]>5700]) #12223 customers


#summary of univariate Analysis
1.> the avg montly emi  of customer is approximately 8437.618
2.> the median of monthly emi of customer is 5700
3.> this column has highly positive skewness
5.>  The maximum montly emi count is 202300, with 1 customers having highest montly income and the purpose of loan is education
6.> Even if higher montly emis  are statistically identified as outliers by box plot
they do not necessarily represent invalid data, as having high montly emi  is possible depending upon loan requested loan provided and montly income



In [ ]:
#univariate analysis on bank_account_vintage_years

display(df["bank_account_vintage_years"].describe())
#mean-> 3.97
#(max,min)-(25,0.5)
#median-2.8
#(q1,q3)->(1.2,5.4)
df["bank_account_vintage_years"].plot(kind="hist")
plt.show()
df["bank_account_vintage_years"].plot(kind="density")
plt.show()
display(df["bank_account_vintage_years"].skew()) #1.906
df["bank_account_vintage_years"].plot(kind="box")
plt.show()



#summary of Analysis
1.> the avg bank account age of customer is 3.97 years
2.> the maximum bank account age is 25 years
3.> the minimum bank account age is 0.5 years
4.> the data is positively skewed
5.> Most customers have a relatively newer bank account
75% of customers have an account vintage of 5.4 years or less

6.> most customers have bank account age in the range 0f 0-5 years
7.> there is a present of outliners after crossing bank account age "10"
after analysing the data we can conclude 
 even after if high bank account age is statically identified as outliners
they do not neccesory represent invalid values



In [ ]:
#univariate analysis on credit_utilizatio_ratio_pct
display(df["credit_utilization_ratio_pct"].describe())
#mean-> 41.54
#(min,max)-(0,100)
#(q1,q3)-(27.7,55)
df["credit_utilization_ratio_pct"].plot(kind="hist",bins=5)
plt.show()
df["credit_utilization_ratio_pct"].plot(kind="density")
plt.show()
display(df["credit_utilization_ratio_pct"].skew())
df["credit_utilization_ratio_pct"].plot(kind="box")
plt.show()

display(df[df["credit_utilization_ratio_pct"]>55].shape)
display(df[df["credit_utilization_ratio_pct"]>65].shape)
display(df[df["credit_utilization_ratio_pct"]>85].shape)
display(df[df["credit_utilization_ratio_pct"]>95].shape)
display(df[df["credit_utilization_ratio_pct"]==100].shape)
a=df[df["credit_utilization_ratio_pct"]==100]
display(a["monthly_income_inr"])
display(a["existing_emi_monthly_inr"])
display(a["existing_loans_count"])
display(a["cibil_score"])


#summary of Analysis
1.> the average Credit Utilization Ratio is 41.54
2.> the minimum Credit Utilization Ratio is 0
3.> the maximum Credit Utilization Ratio is 100
4.> by observing histogram we can say that credit ratio pct of most customer lies btw  40 to 60
5.> the coloum is neglibily skewed (0.118)
6.> by observing box plot we can say that there is outliners in the df but
high credit ratio doesnt mean the data is 100 percent invalid (it's says that the customer may be finanicially bad)
 the customer with high credit ratio pct have bad cibil scores
7.> the median of credit utilization ratio is 41.3




In [ ]:
#univariate analysis on interest_rate_offered_pct
display(df["interest_rate_offered_pct"].describe())
#mean-> 20.71 
#min-> 10.57
#max-> 26
#q1-> 18.81
#q3-> 23.51
df["interest_rate_offered_pct"].plot(kind="hist")
plt.show()
df["interest_rate_offered_pct"].plot(kind="density")
plt.show()
display(df["interest_rate_offered_pct"].skew())
df["interest_rate_offered_pct"].plot(kind="box")
plt.show()


In [ ]:
display(df[df["interest_rate_offered_pct"]<20])
a=df[df["interest_rate_offered_pct"]<20]
display(a["cibil_score"].describe())
display(a["existing_emi_monthly_inr"].describe())
display(a["late_payments_last_12m"].describe())
display(a["existing_loans_count"].describe())
b=a[a["late_payments_last_12m"]==8]
display(b["existing_loans_count"])
display(df[df["interest_rate_offered_pct"]==10.57])

a=df[df["interest_rate_offered_pct"]<15]
display(a["cibil_score"].describe())


#summary of univariate analysis on interest_rate_offered_pct (percent)
1.> the avg interest rate offered is 20.71 
2.> the minimum interest rate offered is 10.57
3.> the maximum interset rate offered is 26
4.> the median interest rate is 21.05
5.> by obseving histogran we can say that no of customers increases in the range of 18 to 26 percent
6.> the column is negatively skewd (-0.47)
7.> by obseving box plot there is outliners before 21 percent interst rate but if a bank is offering low interest that doesnt mean the data is invalid
8.> infact the customer with the min interest rate has a cibil score of 788 with a good income and no existing loan counts
hence the data is valid





In [ ]:
#univariate Analysis on Cibil score coloumn

display(df["cibil_score"].describe())
#1.> the Avg cibil score of a person is 615
#2.> the Median cibil score of a person is 617
#3.> the minimum cibil score is 360 while the max is 866
#4.> (q1,q3)->(574,659)
df["cibil_score"].plot(kind="hist")
plt.show()
df["cibil_score"].plot(kind="density")
plt.show()
display(df["cibil_score"].skew())
df["cibil_score"].plot(kind="box")
plt.show()


In [ ]:
#analaysis of customers having low cibil score
display(df[df["cibil_score"]==360])
#we can clearly see the customer currently have 5 existing loans 
#the customer has late payments 5 times in the past one year
#even interest rate is also higher

a=df[df["cibil_score"]<500]
display(a.describe())
#we can clearly see that the 75 percent of  customers who have a cibil score less than 500 has late paytments 4 times in the past one year
#existing loan count 2 which is quite normal
#as 75% of customers have interset rate of 24 percent due to thier bad cibil score



In [ ]:
#analaysis of customers having high cibil score
display(df[df["cibil_score"]==866])
#we can clearly see the customer currently have 0  existing loans 
#the customer has late payments 0 times in the past one year
#even interest rate is also lowest due to her very good financial profile

a=df[df["cibil_score"]>770]
display(a.describe())
#we can clearly see that the 75 percent of  customers who have a cibil score more  than 750 has late paytments 0 times in the past one year
#existing loan count 1 which is quite normal due to thier good cibil score
#as 75% of customers have interset rate of 13.7 percent due to thier good cibil score



#summary
1.> the Avg cibil score of a person is 615
2.> the Median cibil score of a person is 617
3.> the minimum cibil score is 360
4.> while the max is 866
5.> by looking at the histogram we can say that most  of the cibil scores lies btw the range of approx 600-700
6.> the coloumn is neglibilicy negatively skewed (-0.13)
7.> by looking at the box plot they are outliners below 500 and above 750
but we have analysed the behaviour of customers having a bad cibil scores and as well as good cibil score
8.> by that we can conclude that having a gernally high or low scores doesn't means a invalid value
but it can tell a lot about financial understanding of a particular customer
9.> high cibil score gernally means low interest rate,low late payments,low loan counts and a better financial image
10.> whreas  low cibil score gernally means high interest rate,high  late payments,high  loan counts and a bad  financial image

# A very important result from this Analysis
customers with cibil score <500 have a 42.6% default rate whereas customers with cibil score >770 have a 0% default rate — indicating cibil_score is a strong predictor of default_flag

In [ ]:
#univariate Analysis on num_credit_inquiries_last_6m column
display(df["num_credit_inquiries_last_6m"].describe())
#mean-> 1.297
#median-> 1
#max-> 9
#min-> 0
#(q1,q2)-> (0,2)


df["num_credit_inquiries_last_6m"].plot(kind="box")
plt.show()
#there are outlines above 6 as shown on box plot
a=df[df["num_credit_inquiries_last_6m"]>=6]
display(a.describe())



#summary of univariate analysis on num_credit_inquiries_last_6m

1.> average number of credit inquiries in the last 6 months is around 1.3
2.> most customers have 0 to 2 inquiries, which is normal behavior
3.> a small group of customers (54 people) have 6 or more inquiries, 
which the boxplot flags as outliers
4.> initially it seems like high inquiries should mean higher risk, 
since applying for credit repeatedly can look desperate
5.> but when i checked the default rate of this high-inquiry group, 
it turned out to be quite low (~13%) - much lower than what i expected
6.> this tells me that high inquiries alone don't make someone risky - 
they might just be comparing loan offers or genuinely need credit 
    for multiple valid reasons (like a personal loan)
7.> so unlike cibil_score, this column doesn't seem to be a strong 
standalone predictor of default on its own

In [ ]:
#Univariate Analysis on loan_amount_requested_inr
display(df["loan_amount_requested_inr"].describe())
df["loan_amount_requested_inr"].plot(kind="hist")
plt.show()
#df["loan_amount_requested_inr"].plot(kind="density")
#plt.show()
display(df["loan_amount_requested_inr"].skew())
#df["loan_amount_requested_inr"].plot(kind="box",style="plain")
#plt.show()


#summary of analysis
1.> the Average amount of loan requested is 158493
2.> the median of amount of loan requested is 133000
3.> By seeing the histogram we can say that as loan amt increases no of customer decreases
4.> the coloumn is highly positively skewed (2.303)
5.> there is a very much outliners on barplot but that doesnt mean data is invalid because the loan amt requested logically depend on customer
purpose of loan,montly emi,monthly income
6.> so high loan amt or low loan amt requested is a valid data depending on customer needs


In [ ]:
#univariate analysis on late_payment_last_12m
display(df["late_payments_last_12m"].describe())
df["late_payments_last_12m"].plot(kind="hist")
plt.show()
df["late_payments_last_12m"].plot(kind="density")
plt.show()
display(df["late_payments_last_12m"].skew())
df["late_payments_last_12m"].plot(kind="box")
plt.show()

#summary of analysis
1.> the avg of late payment in last 12 months is 1.44
2.> the median is 1
3.> there is a customer wh0 paid thier montly emi 8 times late in the past 1 year
4.> also there is a customer  who paid thier all monthly emis on time in last 1 year
5.> by observing the histogram we can say that most of customer have atleast 1 late payment in last 1 year
6.> the coloumn is postively skewed (0.9)
7.> there is outlines on the boxplot after "5" late payments in the last 12m
8.> high late payment doesn't neccesirly means invalid data even it helps us to find those customer who have a high default risk score
9.>  75 percent of customers with high late payment has a defalult risk of nearly 70


In [ ]:
display(df[df["late_payments_last_12m"]==8])
#there are 8 customers in the dataset who cant pay thier monthly emis on time (8 times) in the past year
#these customers have a very bad cibil scores due to thier behaviours which is obvious and very high interst rate on thier loans
#even they have a very high chance of becoming a defaulter
#the fact is that 5 out of 8 (more than 60 percent ) have default flag=1
a=df[df["late_payments_last_12m"]>5]
display(a.describe())
display(a.shape)
#there are 149vcustomers in the dataset who cant pay thier monthly emis on time (more than 5 timess) in the past year
#these customers have a relatively  bad cibil scores (75 % percent have cibil score of 529)
#these customers have a relatively high default risk score, with 75% of them having a score of nearly 70


In [ ]:
#Univariate Analysis on loan tenture months column
display(df["loan_tenure_months"].describe())
#the avg of loan tenture months is 42.4
#the median is 36 months
#the minimum  is 12 monthhs
#the maximum is 84 months
display(df["loan_tenure_months"].skew())
df["loan_tenure_months"].plot(kind="box")
plt.show()
#no visible outliners on box plot
display(df[df["loan_tenure_months"]==84])
a=df[df["loan_tenure_months"]==84]
display(a.describe())
#only 6.4 % default_flag longer tenture months = low risk




In [ ]:
#univariate analysis on default_risk_score
display(df["default_risk_score"].describe())
df["default_risk_score"].plot(kind="hist")
plt.show()
df["default_risk_score"].plot(kind="density")
plt.show()
display(df["default_risk_score"].skew())
df["default_risk_score"].plot(kind="box")
plt.show()


In [ ]:
a = df[df["default_risk_score"]>20]
display(a["default_flag"].describe())
display(a.shape)


count    2100.000000
mean        0.375714
std         0.484422
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         1.000000
Name: default_flag, dtype: float64

(2100, 22)

In [25]:
display(df["default_risk_score"].corr(df["default_flag"]))
b = df[df["default_risk_score"]>50]
display(b["default_flag"].describe())
display(b.shape)

np.float64(0.4854031544497407)

count    480.000000
mean       0.664583
std        0.472628
min        0.000000
25%        0.000000
50%        1.000000
75%        1.000000
max        1.000000
Name: default_flag, dtype: float64

(480, 22)

#summary of default_risk_score
1.> the avg default risk score is 6.26 but the median is only 1.77, 
    which shows the column is heavily right skewed
2.> the skewness value is 3.63, which is the highest skew seen in the 
    whole dataset so far
3.> 75 percent of customers have a risk score below 5.85, but the max 
    goes all the way up to 99.08, meaning a small group of customers 
    carry a very high risk score
4.> checked the correlation between default_risk_score and default_flag, 
    it came out to be 0.485, a moderate positive correlation, meaning 
    this score does relate to actual defaults but is not a perfect 
    standalone predictor
5.> when checking by threshold, the default rate increases as the risk 
    score increases:
6.>customers with score > 20 have a 37.6% default rate
7.>customers with score > 50 have a 66.5% default rate
8.>this confirms that default_risk_score is a meaningful indicator of 
actual default risk, even though its not the only factor deciding it